### ISSUE TITLE
Question: nextjs13 server components and 

### ISSUE THREAD

**Issue by droganov** [nanostores/nanostores#218]
I checked if I can use nanostroes with server components in Nextjs, and didn't find any luck.
The below setup renders, but the store init does not fire, and I find store with the initial state both on on server and client.

Did anyone manage to use nanostores with Nextjs server components?

counterStore.ts
```ts
import { useStore } from '@nanostores/react'
import { atom, onMount, task } from 'nanostores'

const store = atom(0)

export const increment: VoidFunction = () => {
  store.set(store.get() + 1)
}

export const decrement: VoidFunction = () => {
  store.set(store.get() - 1)
}

export const initCounter = store.listen(() => {})

export const useCounter = (): number => useStore(store)

onMount(store, () => {
  console.log('onMount: ')
  task(async () => {
    const next = await Promise.resolve(10)
    console.log('next: ', next)
    store.set(next)
  })
})
```

page.tsx
```tsx
import { allTasks } from 'nanostores'

import { Counter } from './Counter'
import { initCounter } from './counterStore'

const NanoPage = async (): Promise<JSX.Element> => {
  initCounter()
  await allTasks()
  return <Counter />
}

export default NanoPage
```

Counter.tsx
```tsx
'use client'
import type { FunctionComponent } from 'react'

import { decrement, increment, useCounter } from './counterStore'

export const Counter: FunctionComponent = () => {
  const count = useCounter()
  console.log('count: ', count)
  return (
    <>
      page {count}
      <hr />
      <button className="yobta-button inline-flex" onClick={decrement}>
        -
      </button>
      <button className="yobta-button inline-flex" onClick={increment}>
        +
      </button>
    </>
  )
}
```

**Comment by ai** [comment]
@dkzlv we need an RSC expert here

**Comment by ai** [comment]
@droganov what is your expected and actual results?

**Comment by droganov** [comment]
I expected mount to fire during SSR, what the state to be 10 after await allTasks()
on client I expected a store to loose the state and initialize with and then update to 10 after mount, because we obviously need to hack next cache to pickup ssr state on client.

I think the issue is related to how nextjs bundler works, looks like it only ues what's explicitly imported, and I'm curious if anyone has used nanostores with RSC. Looks like it's just two different paradgms

**Comment by droganov** [comment]
Looks like readme is incorrect:
https://github.com/nanostores/nanostores#tasks

`store.listen(() => {})` does not fire `mount`, but `subscribe` does

this is the working setup:

```ts
import { useStore } from '@nanostores/react'
import { allTasks, atom, onMount, task } from 'nanostores'

const store = atom(0)

export const increment: VoidFunction = () => {
  store.set(store.get() + 1)
}

export const decrement: VoidFunction = () => {
  store.set(store.get() - 1)
}

export const useCounter = (): number => useStore(store)

export const prefetch = async (): Promise<number> => {
  const unsubscribe = store.subscribe(() => {})
  await allTasks()
  unsubscribe()
  return store.get()
}

onMount(store, () => {
  task(async () => {
    const next = await Promise.resolve(10)
    store.set(next)
  })
})
```

But there is still a gap, as I said above client store does not receive server state and starts with `0`

Also this setup is actually questionable, RSC's are async, which suggests race conditions. The store state may collide when using `allTasks`

**Comment by ai** [comment]
> store.listen(() => {}) does not fire mount, but subscribe does

It is not true. `listen()` fires mount as well.

**Comment by ai** [comment]
> But there is still a gap, as I said above client store does not receive server state and starts with 0

How Next.js “auto-magically” send the current app state to the client?

**Comment by droganov** [comment]
> How Next.js “auto-magically” send the current app state to the client?
I haven't seen the technical paper how it works, they somehow bundle server data, and you can send props from RSC  to RFC, but the `useStore` is not taking default state, I can solve that with a custom hook, no problem

**Comment by ai** [comment]
> I haven't seen the technical paper how it works, they somehow bundle server data, and you can send props from RSC to RFC

It looks like you believe too much if auto-magic. Next.js needs to know where is your store to dump their values and fill it during the hydration process.

Do we have official RSC docs to add this support to Nano Stores (as I know we need to wait until some real docs will come)?

**Comment by droganov** [comment]
> It looks like you believe too much if auto-magic

that's not not true I wrote above I was skeptical to expect nanstores to pickup state without a hacks.

> Do we have official RSC docs to add this support to Nano Stores?

I'm also sceptical here, the store hook should rely on props, like `useStore(store, { initialState })`.
Nextjs uses modified `fetch` and has cache control, which probably means they have some extras not just RSC. It will be hard to magically discover the state.

**Comment by droganov** [comment]
@dkzlv thank you for update, that was a slip indeed

I will be watching how you eventually solve the hydration problem.

for now I'm a bit concerned with current implementation, when stores fire `mount` during SRR, with RSC it would result in collisions.

Anyway looking forward to see the results

**Comment by RnbWd** [comment]
I'm experimenting with this too, using a more simplified store, but this is how I understand what's happening with nextjs 13:

Initializing a store inside a file, outside of a react component, should be considered a server-only store. This store persists on the server between requests, and can be used for rendering server components. However, the data should be serialized into an object before it's passed to a client component. Also, state from the previous request will be there for the next request, so you shouldn't use it for any sensitive or specific user info. You can pass the store between client components, but you shouldn't pass it from a server to a client component. 

If you initialize the store in file, then the client will always start with the initial store value, and whatever actions happened with the store from server components won't have happened on the client. 

The technique that I'm using is to initialize the atom inside of a useRef, and accessing the atom on the client like this:

```
export const StoreContext = createContext(undefined);

export const useAtom = () =>
  useStore(useContext(StoreContext));
  
export default function StoreProvider({ serverProps, children }) {
  const storeRef = useRef(atom(serverProps));
  console.log('using nanostores');
  useEffect(() => {
    ... client logic
   if (storeRef.current) {
    storeRef.current.set('');
   }
  }, []);

  return (
    <StoreContext.Provider value={storeRef.current}>
      {children}
    </StoreContext.Provider>
  );
}
```

Nextjs is still rendering client components on the server and hydrating properly, because the `useStore` hook has a SSR getter via `useSyncExternalStore`.

I think if you need to use nanostores in client components, then it should only be imported in client components, or initially memoized inside of render functions in client components. I think if you want to `await allTasks()` inside of an async server component, then that nanostore should be server only store that's not shared with client components or accessed inside of client components directly.

**Comment by RnbWd** [comment]
I just looked at the code for `tasks`, and I'm pretty sure that these variables will persist on the server between requests:

```
let tasks = 0
let resolves = []
```
I haven't tested this yet, but I think if you are waiting for tasks from 1 user, when next user sends a request, it will just add to that list and all users requesting nextjs will be awaiting all the tasks for all users.

Right now I'm not doing anything complex with server nanostores, but everything in nextjs is rendered on server first even in client components unless you import a component dynamically or put logic inside of useEffect

**Comment by dkzlv** [comment]
@RnbWd Yes, you are correct.
I have [a draft PR](https://github.com/nanostores/nanostores/pull/228) that introduces Contexts, and those module-scope variables are moved into contexts, that are supposed to be user/request specific, so there shouldn't be any conflicts. As we said, we're not really SSR-ready right now, but we really hope to become in short-term future, maybe a couple of weeks.

**Comment by bernatfortet** [comment]
Any updates on this end?

**Comment by ai** [comment]
@bernatfortet go to PR and see where help is needed
